src/org/rascalmpl/library/lang/html/IO.java

```diff
@@ -12,37 +12,63 @@
 package org.rascalmpl.library.lang.html;
 
 import java.io.IOException;
+import java.io.OutputStream;
 import java.io.Reader;
 import java.io.StringReader;
+import java.io.StringWriter;
 import java.net.MalformedURLException;
 import java.util.ArrayList;
 import java.util.Enumeration;
 import java.util.HashMap;
+import java.util.Map;
+import java.util.Map.Entry;
+import java.util.NoSuchElementException;
 import java.util.Stack;
 
 import javax.swing.text.MutableAttributeSet;
 import javax.swing.text.html.HTML.Tag;
 import javax.swing.text.html.HTMLEditorKit.ParserCallback;
 import javax.swing.text.html.parser.ParserDelegator;
+import javax.xml.parsers.DocumentBuilder;
+import javax.xml.parsers.DocumentBuilderFactory;
+import javax.xml.parsers.ParserConfigurationException;
+import javax.xml.transform.Transformer;
+import javax.xml.transform.TransformerException;
+import javax.xml.transform.TransformerFactory;
+import javax.xml.transform.dom.DOMSource;
+import javax.xml.transform.stream.StreamResult;
 
 import org.rascalmpl.exceptions.RuntimeExceptionFactory;
 import org.rascalmpl.uri.URIResolverRegistry;
+import org.w3c.dom.Document;
+import org.w3c.dom.Element;
+import org.w3c.dom.Node;
 
-import io.usethesource.vallang.INode;
+import io.usethesource.vallang.IConstructor;
+import io.usethesource.vallang.IList;
 import io.usethesource.vallang.ISourceLocation;
 import io.usethesource.vallang.IString;
 import io.usethesource.vallang.IValue;
 import io.usethesource.vallang.IValueFactory;
+import io.usethesource.vallang.type.Type;
+import io.usethesource.vallang.type.TypeStore;
 
 public class IO {
     private final IValueFactory factory;
+    private final TypeStore store;
+    private final Type HTMLElement;
 
-    public IO(IValueFactory factory) {
+    public IO(IValueFactory factory, TypeStore store) {
         this.factory = factory;
+        this.store = store;
+        this.HTMLElement = store.lookupAbstractDataType("HTMLElement");
     }
     
-
     public IValue readHTMLString(IString string) {
+        if (string.length() == 0) {
```

Comment:
We should really think about adding `isEmpty` to vallang IString, especially since length is not always a o(1) operation.

Reply:
Well it's log(n) max with the current balanced tree implementation, so you'd have to call it many times on a deep tree before you would notice any improvement. It's doable though to add isEmpty with a default false implementation in the interface and only return true for the maximally shared empty constant.

Reply:
Also length is cached in internal nodes I believe. So it's probably even in O(1) already.

Reply:
I think a default method should be: return length() == 0; and then subclasses can override it with true or false.

Reply:
Let's Have a look at IStringImplementation first. It's a private interface used by all implementation classes of IString, all of them but one are non-empty I believe. And that one is empty.

Reply:
I'm just saying the default interface method should work for future overloads as well.